server/src/main/scala/com/cloudera/livy/server/interactive/InteractiveSession.scala

```diff
@@ -376,6 +376,9 @@ class InteractiveSession(
     }
     uriFuture onFailure { case e => warn("Fail to get rsc uri", e) }
 
+    // Register this class to RSCClient as a session state listener
+    client.get.registerStateListener(this)
```

Comment:
But it's a race condition that could happen. I wouldn't introduce a race condition to simplify something.
Originally I took a similar approach too and moved away from it after spotting this race condition.